### DISCUSSION TITLE
Is it possible to execute tasks in the background using chat prompts and show progress in the chat window.

### DISCUSSION THREAD

**Discussion by kamalkthakur** [open-webui/open-webui#3672] [Q&A]
I am new to open-webui. I am looking for an open source GUI tool which can execute tasks in the background using chat prompts and show progress in the chat window itself time to time. 
I found two concepts in open-webui known as tools and functions. They can help execute custom tasks but do not provide real time update of the execution.

Is there any way with which this can be achieved. Can this be taken as an enhancement in the open-webui if do not exist already.

Sharing one example of expected timely response based on prompt.
<img width="1025" alt="Screenshot 2024-07-06 at 2 28 40 PM" src="https://github.com/open-webui/open-webui/assets/22568429/0929855b-dc41-467f-8b09-437e9d3deb1c">

**Comment by PlebeiusGaragicus** [comment]
I am also interested in the same thing!  In my use case, I need an Open WebUI "pipeline" that will call my self-hosted LangGraph workflows.  Since these are often complicated graphs, I'd like feedback **during** execution and be able to steam partial results.

I just found a (maybe undocumented) possible feature in this [posted function](https://openwebui.com/t/iamg30/user_info_retrieval_tool/)

You can see:
```python
async def get_user_info(
        self, __event_emitter__: Callable[[dict], Awaitable[None]], __user__: dict = {}
    ) -> str:
```

**__event_emitter__** as a passed callable for the tool.  Just a few weeks ago I was building a UI from scratch that had just this functionality - callbacks inside my models.

I can't find any more examples and this code didn't work for me.  I know that the "pipelines" are still very new and under development, but I am SO hoping that we can get something akin to this similar to what the Chainlit library provides for their chat interface.

**Reply by kamalkthakur** [reply]
I recently found AgentLite and StreamLit. You may find them useful. UI is not clean. But really serves the purpose.

https://github.com/SalesforceAIResearch/AgentLite
https://github.com/streamlit/streamlit

**Reply by Micropot** [reply]
@PlebeiusGaragicus Have you find a way to use event_emitter into a pipeline ? I'm intreasted about doing the same but i cannot figure out how to implement it

**Reply by PlebeiusGaragicus** [reply]
No - I haven't found a way to do this yet.  I understand I can do it with a tool/function but that doesn't meet my needs as I'm using PIPELINES.

Also, yes I've used Streamlit to build my own ChatGPT clone BEFORE I found Open WebUI.  My goal is to spend time building a collection of useful agents... not to re-invent the wheel and build my own UI.  I was spending 90% of my time building my Streamlit chat web interface and now that I'm using Open WebUI I can just focus on building agents.

I'm using LangChain/LangGraph and it would be a KILLER feature to see which node is running during computation.

**Comment by smonux** [comment]
Tools can send __event_emitters__ messages . The "status" one is done specifically for this purpose, but you can also send messages which go to the llm output.

https://docs.openwebui.com/features/plugin/tools/

![imagen](https://github.com/user-attachments/assets/6514288e-e22a-4df7-8af2-238c03917bb2)

**Reply by Micropot** [reply]
i know, but i don't want to use tools. I'm creating a pipeline to chat with my chatbot, and creating multiple Events using Llama_Index. I dont think it's possible to force using a tool everytime a user is asking a question to the chatbot, whithout clicing on a button. Maybe i'm wrong ;)
